stages/org.osbuild.zipl-secex.inst

```diff
@@ -0,0 +1,234 @@
+                      "-k", hostkey,
+                      "-o", sdboot],
+                     check=True)
+
+
+def open_dmverity(blocksize, partlabel, datapart, hashpart, mountpoint):
+    hash = subprocess.run(["/usr/sbin/veritysetup",
+                            "format", datapart, hashpart,
+                            "--data-block-size", f"{blocksize}"],
+                           check=True, capture_output=True)
+    for line in hash.stdout.decode().splitlines():
+      if line.startswith("Root hash:"):
+        hash = line.split(":", 1)[1].strip()
+        break
+
+    subprocess.run(["/usr/sbin/veritysetup",
+                    "open", datapart, partlabel, hashpart, hash],
+                   check=True)
+    subprocess.run(["mount", "-o", "ro", f"/dev/mapper/{partlabel}", mountpoint],
+                   check=True)
+    return hash
+
+
+def close_dmverity(partlabel, mountpoint):
+    subprocess.run(["umount", mountpoint], check=True)
+    subprocess.run(["/usr/sbin/veritysetup", "close", partlabel], check=True)
+
+
+def main(tree, paths, devices, options):
+    kver = options["kernel"]
+    append_kopts = options.get("kernel_opts_append", [])
+    blocksize = options.get("blocksize", 512)
+    build_id = options["build_id"]
+    location = options["location"]
+
+    hostkey = os.path.join(tree, options["hostkey"])
+    public_key = os.path.join(tree, options["ignition_gpg_pubkey"])
+
+    disk = os.path.join(paths["devices"], devices["disk"]["path"])
+    target = f"{paths["mounts"]}/se"
+
+    boot_fs = f"{disk}p3"
+    boothash_fs = f"{disk}p5"
+    root_fs = f"{disk}p4"
+    roothash_fs = f"{disk}p6"
+
+    with tempfile.TemporaryDirectory() as sysroot:
+      try:
+        roothash = open_dmverity(blocksize, "root", root_fs, roothash_fs, sysroot)
+        boothash = open_dmverity(blocksize, "boot", boot_fs, boothash_fs, f"{sysroot}/boot")
+
+        kernel, initrd, kopts = find_kernel(sysroot, kver)
+
+        secret_key = generate_gpg_keys(public_key, build_id)
+        sdboot = os.path.join(target, "sdboot")
+        with append_to_initrd(initrd, "/usr/lib/coreos/ignition.asc", secret_key) as initrd:
+          if append_kopts:
+            kopts = f"{kopts} {' '.join(append_kopts)}"
+          kopts = f"{kopts} rootfs.roothash={roothash}"
+          kopts = f"{kopts} bootfs.roothash={boothash}"
+          genprotimg(kernel, initrd.name, kopts, hostkey, sdboot)
+
+        subprocess.run(["/usr/sbin/zipl",
+                        "--verbose",
+                        "--target", target,
+                        "--image", sdboot,
+                        "--targetbase", disk,
+                        "--targettype", "SCSI",
+                        "--targetblocksize", str(blocksize),
+                        "--targetoffset", str(location)],
+                      check=True)
+      finally:
+        close_dmverity("boot", f"{sysroot}/boot")
+        close_dmverity("root", sysroot)
```

Comment:
if a failure occurred in `open_dmverity(blocksize, "root", root_fs, roothash_fs, sysroot)` then we would have never set up `boot` and this would fail since `check=True` inside `close_dmverity()` right? Then `close_dmverity("root", sysroot)` wouldn't get called?

Is there a way we can do a best effort here?

Reply:
i'm no expert in python, but a second, nested try-finally should work and be ugly?

Reply:
I think I would just use `check=False` for these two (best effort)